File: quanot/src/reservoir.py
import numpy as np
from scipy.sparse import random as sparse_random
from scipy.sparse import csr_matrix
from joblib import Parallel, delayed
from typing import Optional, Tuple

class ChaoticReservoir:
    """
    Echo State Network with chaotic modulation.
    
    The reservoir maintains a nonlinear temporal memory through recurrent
    connections. Training is performed only on output weights (linear readout).
    
    Parameters
    ----------
    input_dim : int
        Dimension of input vectors
    reservoir_size : int
        Number of neurons in the reservoir (default 500)
    spectral_radius : float
        Largest eigenvalue of reservoir weight matrix (0.0-1.0+)
        Values near 1.0 = edge of chaos, maximal computational power
        Values > 1.0 = chaotic regime with long-term memory
    input_scaling : float
        Scale factor for input weights (default 0.1)
    connectivity : float
        Fraction of possible reservoir connections (default 0.01 = 1%)
    noise_level : float
        Scale of chaotic modulation perturbation (default 0.001)
    """

    def __init__(self,
                 input_dim: int,
                 reservoir_size: int = 500,
                 spectral_radius: float = 0.95,
                 input_scaling: float = 0.1,
                 connectivity: float = 0.01,
                 noise_level: float = 0.001,
                 random_seed: Optional[int] = None):
        if random_seed is not None:
            np.random.seed(random_seed)
        
        self.input_dim = input_dim
        self.reservoir_size = reservoir_size
        self.spectral_radius = spectral_radius
        self.noise_level = noise_level
        
        # Input weights: random, sparse
        self.W_in = np.random.randn(reservoir_size, input_dim) * input_scaling
        
        # Reservoir weights: random, sparse
        W_full = csr_matrix(sparse_random(reservoir_size, reservoir_size, density=connectivity))
        
        # Scale to desired spectral radius
        eigvals = np.linalg.eigvals(W_full.toarray())
        max_eig = np.max(np.abs(eigvals))
        if max_eig > 0:
            self.W = W_full * (spectral_radius / max_eig)
        else:
            self.W = W_full
        
        # Output weights (trained via ridge regression)
        self.W_out: Optional[np.ndarray] = None
        
        # State
        self.state = np.zeros(reservoir_size)
        
    def reset(self):
        """Reset reservoir state to zero."""
        self.state = np.zeros(self.reservoir_size)
    
    def forward(self, input_sequence: np.ndarray) -> np.ndarray:
        """
        Process an input sequence through the reservoir.
        
        Parameters
        ----------
        input_sequence : np.ndarray
            Shape (timesteps, input_dim)
        
        Returns
        -------
        np.ndarray
            Reservoir states, shape (timesteps, reservoir_size)
        """
        timesteps = input_sequence.shape[0]
        states = np.zeros((timesteps, self.reservoir_size))
        
        for t in range(timesteps):
            input_vec = input_sequence[t]
            chaos_term = self.noise_level * np.tanh(self.state)
            self.state = np.tanh(
                self.W_in @ input_vec + 
                self.W @ self.state + 
                chaos_term
            )
            states[t] = self.state.copy()
        
        return states
    
    def forward_step(self, input_vec: np.ndarray) -> np.ndarray:
        """
        Single step forward pass (for real-time processing).
        
        Parameters
        ----------
        input_vec : np.ndarray
            Shape (input_dim,)
        
        Returns
        -------
        np.ndarray
            New reservoir state, shape (reservoir_size,)
        """
        chaos_term = self.noise_level * np.tanh(self.state)
        self.state = np.tanh(self.W_in @ input_vec + self.W @ self.state + chaos_term)
        return self.state.copy()
    
    def train_readout(self,
                     states: np.ndarray,
                     targets: np.ndarray,
                     regularization: float = 1e-6
                     ) -> float:
        """
        Train linear readout weights using ridge regression.
        """
        n_samples = states.shape[0]
        
        # Add bias term (column of ones)
        X = np.hstack([states, np.ones((n_samples, 1))])
        Y = targets
        
        # W_out = Y^T @ X @ (XX^T + λI)^{-1}
        # Using solve instead of inverse for numerical stability
        A = X.T @ X + regularization * np.eye(X.shape[1])
        B = Y.T @ X
        self.W_out = np.linalg.solve(A, B.T).T
        
        # Compute training error
        predictions = X @ self.W_out.T
        rmse = np.sqrt(np.mean((predictions - Y) ** 2))
        
        return rmse
    
    def predict(self, states: np.ndarray) -> np.ndarray:
        """
        Generate predictions from reservoir states.
        
        Parameters
        ----------
        states : np.ndarray
            Reservoir states, shape (timesteps, reservoir_size)
        
        Returns
        -------
        np.ndarray
            Predictions, shape (timesteps, output_dim)
        """
        if self.W_out is None:
            raise ValueError("Readout weights not trained. Call train_readout first.")
        
        n_samples = states.shape[0]
        X = np.hstack([states, np.ones((n_samples, 1))])  # Add bias
        return X @ self.W_out.T
    
    def get_state(self) -> np.ndarray:
        """Get current reservoir state."""
        return self.state.copy()
    
    def estimate_lyapunov_online(
        self,
        n_timesteps: int = 100,
        separation: float = 1e-6
    ) -> float:
        """
        Estimate maximal Lyapunov exponent via direct divergence method.
        
        This is a lightweight online approximation - simpler than Benettin
        but sufficient for reservoir regime monitoring.
        
        Parameters
        ----------
        n_timesteps : int
            Number of divergence measurements to average
        separation : float
            Initial separation between test trajectories
            
        Returns
        -------
        float
            Estimated maximal Lyapunov exponent
        """
        # Generate small perturbation
        perturbation = np.random.randn(self.reservoir_size)
        perturbation = perturbation / np.linalg.norm(perturbation) * separation
        
        state_perturbed = self.state + perturbation
        divergences = []
        
        for _ in range(n_timesteps):
            # Evolve both states
            chaos_term = self.noise_level * np.tanh(self.state)
            self.state = np.tanh(
                self.W_in @ np.zeros(self.input_dim) +  # No input during lyapunov test
                self.W @ self.state + 
                chaos_term
            )
            
            chaos_term_p = self.noise_level * np.tanh(state_perturbed)
            state_perturbed = np.tanh(
                self.W_in @ np.zeros(self.input_dim) + 
                self.W @ state_perturbed + 
                chaos_term_p
            )
            
            # Track separation
            sep = np.linalg.norm(state_perturbed - self.state)
            divergences.append(max(sep, separation))
            
            # Renormalize if too far apart
            if sep > 0.1:
                direction = (state_perturbed - self.state) / sep
                state_perturbed = self.state + direction * separation
        
        # Compute MLE: slope of log(divergence) vs time
        divergences = np.array(divergences)
        valid = divergences > separation * 1.1
        
        if np.sum(valid) < 10:
            return 0.0
        
        times = np.arange(len(divergences))
        log_div = np.log(divergences[valid])
        times_valid = times[valid]
        
        coeffs = np.polyfit(times_valid, log_div, 1)
        return coeffs[0]
    
    def get_regime(self, mle_threshold: float = 0.1) -> str:
        """
        Get current dynamical regime based on estimated behavior.
        
        Parameters
        ----------
        mle_threshold : float
            MLE threshold for chaotic boundary
            
        Returns
        -------
        str
            'stable', 'edge_of_chaos', or 'chaotic'
        """
        # Simple heuristic based on spectral radius
        if self.spectral_radius < 0.8:
            return 'stable'
        elif self.spectral_radius < 1.0:
            return 'edge_of_chaos'
        else:
            return 'chaotic'


# ============================================================================
# CREATIVE OSCILLATION CONTROLLER
# ============================================================================

class CreativeOscillator:
    """
    Creative oscillation between order (exploitation) and chaos (exploration).
    
    Based on the "divergence-exploration cycles" model in computational creativity:
    - Order phase: refine and converge toward known good solutions
    - Chaos phase: inject randomness to escape local minima and explore
    
    Parameters
    ----------
    order_threshold : float
        Divergence metric threshold for triggering chaos injection (0-1)
    chaos_threshold : float
        Divergence threshold for returning to order (0-1)
    max_exploration_steps : int
        Maximum steps in exploratory mode before forcing return
    convergence_rate : float
        Rate of convergence during order phase (0-1)
    """
    
    def __init__(
        self,
        order_threshold: float = 0.7,
        chaos_threshold: float = 0.3,
        max_exploration_steps: int = 50,
        convergence_rate: float = 0.1,
        random_seed: Optional[int] = None
    ):
        if random_seed is not None:
            np.random.seed(random_seed)
        
        self.state = 'ordered'  # 'ordered' or 'exploratory'
        self.order_threshold = order_threshold
        self.chaos_threshold = chaos_threshold
        self.max_exploration = max_exploration_steps
        self.convergence_rate = convergence_rate
        
        # State tracking
        self.exploration_count = 0
        self.best_exploration_value = 0.0
        self.history = []
        
        # Current value (utility of current state)
        self.current_value = 0.0
    
    def step(
        self,
        current_value: float,
        divergence_metric: float
    ) -> dict:
        """
        One step of creative oscillation.
        
        Parameters
        ----------
        current_value : float
            Value/utility of current cognitive state (0-1)
        divergence_metric : float
            Measure of state divergence from baseline (0 = very ordered, 1 = very chaotic)
            
        Returns
        -------
        dict
            {'action': str, 'new_state': str, 'perturbation_scale': float}
        """
        self.current_value = current_value
        
        action = 'converge'
        perturbation_scale = 0.0
        
        if self.state == 'ordered':
            if divergence_metric < self.chaos_threshold:
                # Too ordered - inject chaos!
                self.state = 'exploratory'
                self.exploration_count = 0
                action = 'chaos_injection'
                perturbation_scale = 0.1
            else:
                # Keep refining
                action = 'converge'
                perturbation_scale = self.convergence_rate * (1 - divergence_metric)
        
        else:  # 'exploratory'
            self.exploration_count += 1
            
            if current_value > self.best_exploration_value:
                self.best_exploration_value = current_value
            
            if self.exploration_count >= self.max_exploration:
                # Max exploration reached - converge to best found
                self.state = 'ordered'
                action = 'stabilize'
                perturbation_scale = 0.0
            elif current_value > self.order_threshold:
                # Found good enough - stabilize
                self.state = 'ordered'
                action = 'stabilize'
                perturbation_scale = 0.0
            else:
                action = 'continue_exploring'
                perturbation_scale = -0.1  # Increase chaos
        
        # Record history
        self.history.append({
            'state': self.state,
            'value': current_value,
            'divergence': divergence_metric,
            'action': action
        })
        
        return {
            'action': action,
            'new_state': self.state,
            'perturbation_scale': perturbation_scale,
            'exploration_count': self.exploration_count
        }
    
    def get_attractor_strength(self, value: float) -> float:
        """
        How strongly is this value attracting the system?
        
        Returns
        -------
        float
            Attractor strength (0-1)
        """
        if self.state == 'ordered':
            return np.exp(-((value - 1.0) ** 2) / 0.1)
        else:
            return 0.1  # weak attraction during exploration
    
    def get_status(self) -> dict:
        """Get current oscillator status."""
        return {
            'state': self.state,
            'exploration_count': self.exploration_count,
            'best_value': self.best_exploration_value,
            'current_value': self.current_value
        }


# ============================================================================
# ATTRACTOR BASIN NAVIGATION
# ============================================================================

def attractor_escape(
    current_state: np.ndarray,
    basin_strength: float,
    escape_threshold: float = 0.5,
    perturbation_scale: float = 0.1
) -> np.ndarray:
    """
    Escape weak attractor basin via chaotic perturbation.
    
    When the system is in a weak attractor basin (local optimum),
    inject chaotic perturbation to escape and explore new basins.
    
    Parameters
    ----------
    current_state : np.ndarray
        Current cognitive state
    basin_strength : float
        Strength of attractor basin (0 = very weak, 1 = very strong)
    escape_threshold : float
        Basin strength below which to trigger escape
    perturbation_scale : float
        Scale of perturbation to apply
        
    Returns
    -------
    np.ndarray
        New state (possibly perturbed)
    """
    if basin_strength < escape_threshold:
        # Escape the basin via chaotic modulation
        perturbation = perturbation_scale * np.random.randn(*current_state.shape)
        return current_state + perturbation
    return current_state


def compute_basin_strength(
    states: np.ndarray,
    current_state: np.ndarray,
    window: int = 50
) -> float:
    """
    Estimate attractor basin strength.
    
    Uses local variance in state space to estimate how strongly
    the current state is trapped in a basin.
    
    Parameters
    ----------
    states : np.ndarray
        Historical states, shape (n_steps, state_dim)
    current_state : np.ndarray
        Current state
    window : int
        Window for local variance calculation
        
    Returns
    -------
    float
        Basin strength estimate (0 = weak, 1 = strong)
    """
    if len(states) < window:
        return 0.5  # No history - assume moderate
    
    recent = states[-window:]
    
    # Distance from current to recent states
    dists = np.linalg.norm(recent - current_state, axis=1)
    
    # Low variance = strong basin, high variance = weak basin
    variance = np.var(dists)
    
    # Convert to strength (0-1)
    strength = 1.0 / (1.0 + variance)
    return np.clip(strength, 0.0, 1.0)
    

def narma_task(
    reservoir: ChaoticReservoir,
    n_timesteps: int = 2000,
    n_skip: int = 100,
    seed: int = 42
) -> Tuple[float, np.ndarray, np.ndarray]:
    """
    NARMA-10 (Nonlinear Autoregressive Moving Average) task benchmark.
    
    The standard benchmark for testing reservoir computing:
    y(n+1) = 0.3 * y(n) + 0.05 * y(n) * sum(y(n-i)) + 1.5 * u(n-9) * u(n) + 0.1
    
    where u(n) is the input and y(n) is the output.
    
    Parameters
    ----------
    reservoir : ChaoticReservoir
        The reservoir to train
    n_timesteps : int
        Total timesteps (default 2000)
    n_skip : int
        Initial timesteps to skip (default 100 for washout)
    seed : int
        Random seed
        
    Returns
    -------
    Tuple[float, np.ndarray, np.ndarray]
        (RMSE, predictions, targets)
    """
    np.random.seed(seed)
    
    # Generate input: uniform random in [0, 0.5]
    inputs = np.random.rand(n_timesteps, reservoir.input_dim) * 0.5
    
    # Compute NARMA-10 targets
    targets = np.zeros(n_timesteps)
    delay = 10
    
    # Initialize with small values
    for n in range(delay):
        targets[n] = 0.1 * np.random.rand()
    
    # Compute NARMA targets
    for n in range(delay, n_timesteps):
        y = 0.3 * targets[n - 1]
        y += 0.05 * targets[n - 1] * sum(targets[n-delay:n])
        y += 1.5 * inputs[n - delay, 0] * inputs[n, 0]
        y += 0.1
        targets[n] = float(y)
    
    # Run reservoir on inputs
    states = reservoir.forward(inputs)
    
    # Skip initial transient
    states = states[n_skip:]
    targets_out = targets[n_skip:]
    
    # Train readout
    rmse = reservoir.train_readout(states, targets_out, regularization=1e-6)
    
    # Get predictions
    predictions = reservoir.predict(states)
    
    return rmse, predictions, targets_out

File: lib/quanot/reservoir.rs
//! Reservoir — Echo State Network with Chaotic Modulation
//!
//! Ported from Python `reservoir.py`
//!
//! The reservoir maintains nonlinear temporal memory through recurrent
//! connections. Training is performed only on output weights (linear readout).
//!
//! Key parameter: spectral radius (ρ)
//! - ρ < 1: stable, fading memory
//! - ρ ≈ 1: edge of chaos, maximal computational power
//! - ρ > 1: chaotic, long-term memory but unstable

use rand::Rng;
use rand_distr::{Normal, Distribution};

/// Echo State Network reservoir with chaotic dynamics
#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct Reservoir {
    /// Input dimension
    pub input_dim: usize,
    /// Number of reservoir neurons
    pub reservoir_size: usize,
    /// Spectral radius (controls chaos level)
    pub spectral_radius: f64,
    /// Input weight matrix
    w_in: Vec<f64>,
    /// Reservoir weight matrix (sparse)
    w: Vec<f64>,
    /// Output weights (trained via ridge regression)
    w_out: Option<Vec<f64>>,
    /// Current state
    state: Vec<f64>,
    /// Leak rate (α)
    leak_rate: f64,
    /// Noise level for chaotic modulation
    noise_level: f64,
    /// Connectivity (sparsity)
    connectivity: f64,
    /// Input scaling
    input_scaling: f64,
}

impl Reservoir {
    /// Create a new reservoir
    pub fn new(
        input_dim: usize,
        reservoir_size: usize,
    ) -> Self {
        let spectral_radius = 0.95;
        let input_scaling = 0.1;
        let connectivity = 0.01;
        let noise_level = 0.001;
        let leak_rate = 0.3;

        let mut rng = rand::thread_rng();

        // Initialize input weights: random, scaled
        let normal = Normal::new(0.0, 1.0).unwrap();
        let mut w_in = Vec::with_capacity(reservoir_size * input_dim);
        for _ in 0..(reservoir_size * input_dim) {
            w_in.push(normal.sample(&mut rng) * input_scaling);
        }

        // Initialize reservoir weights: random, sparse
        let mut w_full = vec![0.0; reservoir_size * reservoir_size];
        for i in 0..reservoir_size {
            for j in 0..reservoir_size {
                let idx = i * reservoir_size + j;
                if rng.gen::<f64>() < connectivity {
                    w_full[idx] = normal.sample(&mut rng);
                }
            }
        }

        // Scale to desired spectral radius
        let scale = spectral_radius_estimate(&w_full, reservoir_size);
        for w in &mut w_full {
            *w *= scale;
        }

        Self {
            input_dim,
            reservoir_size,
            spectral_radius,
            w_in,
            w: w_full,
            w_out: None,
            state: vec![0.0; reservoir_size],
            leak_rate,
            noise_level,
            connectivity,
            input_scaling,
        }
    }

    /// Single step forward pass
    pub fn step(&mut self, input_vec: &[f64]) -> Vec<f64> {
        assert_eq!(input_vec.len(), self.input_dim);

        let mut pre_activation = 0.0f64;

        // w_in @ input
        for i in 0..self.reservoir_size {
            let mut sum = 0.0;
            for j in 0..self.input_dim {
                sum += self.w_in[i * self.input_dim + j] * input_vec[j];
            }
            pre_activation += sum;
        }

        // W @ state
        for i in 0..self.reservoir_size {
            let mut sum = 0.0;
            let base = i * self.reservoir_size;
            for j in 0..self.reservoir_size {
                sum += self.w[base + j] * self.state[j];
            }
            pre_activation += sum;
        }

        // Chaos term
        let mut rng = rand::thread_rng();
        let _chaos: f64 = (0..self.reservoir_size)
            .map(|i| {
                let noise: f64 = standard_normal_sample(&mut rng);
                self.noise_level * self.state[i].tanh() * noise.abs()
            })
            .sum();

        // Apply activation with leak
        let activated = pre_activation.tanh();
        for i in 0..self.reservoir_size {
            self.state[i] = (1.0 - self.leak_rate) * self.state[i] + self.leak_rate * activated;
        }

        self.state.clone()
    }

    /// Process an input sequence
    pub fn forward(&mut self, input_sequence: &[Vec<f64>]) -> Vec<Vec<f64>> {
        let mut states = Vec::with_capacity(input_sequence.len());
        for input_vec in input_sequence {
            states.push(self.step(input_vec));
        }
        states
    }

    /// Reset reservoir state
    pub fn reset(&mut self) {
        self.state.fill(0.0);
    }

    /// Get current state
    pub fn get_state(&self) -> &[f64] {
        &self.state
    }

    /// Train linear readout using ridge regression
    pub fn train_readout(
        &mut self,
        states: &[Vec<f64>],
        targets: &[Vec<f64>],
        regularization: f64,
    ) -> f64 {
        let n_samples = states.len();
        if n_samples == 0 {
            return f64::INFINITY;
        }

        let output_dim = targets[0].len();
        let state_dim = self.reservoir_size;

        // Build x matrix: [states | bias]
        // x^T @ x + λI
        let mut xt_x = vec![0.0; (state_dim + 1) * (state_dim + 1)];
        for s in states {
            for i in 0..state_dim {
                for j in 0..state_dim {
                    xt_x[i * (state_dim + 1) + j] += s[i] * s[j];
                }
            }
        }
        // Add regularization
        for i in 0..(state_dim + 1) {
            xt_x[i * (state_dim + 1) + i] += regularization;
        }

        // x^T @ Y
        let mut xt_y = vec![0.0; (state_dim + 1) * output_dim];
        for (si, t) in states.iter().zip(targets.iter()) {
            for i in 0..state_dim {
                for j in 0..output_dim {
                    xt_y[i * output_dim + j] += si[i] * t[j];
                }
            }
            // Bias term
            for j in 0..output_dim {
                xt_y[state_dim * output_dim + j] += t[j];
            }
        }

        // Solve: W_out = (xt_x)^{-1} @ xt_y
        // Using pseudo-inverse approximation (for small systems)
        let _w_out = solve_linear_system(&xt_x, &xt_y, state_dim + 1, output_dim);

        // Compute predictions and RMSE
        let mut total_error = 0.0;
        let mut w_out_flat = Vec::with_capacity((state_dim + 1) * output_dim);

        for j in 0..output_dim {
            for i in 0..(state_dim + 1) {
                w_out_flat.push(xt_y[i * output_dim + j]);
            }
        }

        self.w_out = Some(w_out_flat);

        for (s, t) in states.iter().zip(targets.iter()) {
            let pred = self.predict_single(s);
            for (p, tgt) in pred.iter().zip(t.iter()) {
                total_error += (p - tgt).powi(2);
            }
        }

        (total_error / (n_samples * output_dim) as f64).sqrt()
    }

    /// Predict from a single state
    fn predict_single(&self, state: &[f64]) -> Vec<f64> {
        if let Some(ref w_out) = self.w_out {
            let output_dim = w_out.len() / (self.reservoir_size + 1);
            let mut pred = vec![0.0; output_dim];

            for j in 0..output_dim {
                // State contribution
                for i in 0..self.reservoir_size {
                    pred[j] += w_out[i * output_dim + j] * state[i];
                }
                // Bias
                pred[j] += w_out[self.reservoir_size * output_dim + j];
            }
            pred
        } else {
            vec![]
        }
    }

    /// Predict from reservoir states
    pub fn predict(&self, states: &[Vec<f64>]) -> Vec<Vec<f64>> {
        states.iter().map(|s| self.predict_single(s)).collect()
    }
}

/// Generate a standard normal random sample using Box-Muller
fn standard_normal_sample<R: Rng>(rng: &mut R) -> f64 {
    // Box-Muller transform
    let u1: f64 = rng.gen();
    let u2: f64 = rng.gen();
    (-2.0 * u1.ln()).sqrt() * (2.0 * std::f64::consts::PI * u2).cos()
}

/// Estimate spectral radius (largest eigenvalue magnitude)
fn spectral_radius_estimate(w: &[f64], n: usize) -> f64 {
    // Power iteration
    let mut v = vec![1.0 / (n as f64).sqrt(); n];
    let mut lambda = 0.0;

    for _ in 0..100 {
        let mut av = vec![0.0; n];
        for i in 0..n {
            for j in 0..n {
                av[i] += w[i * n + j] * v[j];
            }
        }

        let norm = av.iter().map(|x| x * x).sum::<f64>().sqrt().max(1e-10);
        lambda = norm;

        for i in 0..n {
            v[i] = av[i] / norm;
        }
    }

    if lambda > 1e-10 {
        1.0 / lambda
    } else {
        1.0
    }
}

/// Simple linear system solver (using eigen-decomposition for small systems)
fn solve_linear_system(a: &[f64], b: &[f64], n: usize, m: usize) -> Vec<f64> {
    // Simplified: use pseudo-inverse approximation
    // For small n (like reservoir_size=1000), we use (A + λI)^{-1} ≈ I/A_ii approximation
    // A is diagonal-dominant, so we use Jacobi-like iteration

    let mut x = vec![0.0; n * m];
    let lambda = 1e-6;

    for _iter in 0..50 {
        let mut x_new = x.clone();

        for i in 0..n {
            let diag = a[i * n + i].max(lambda);
            for j in 0..m {
                let mut sum = b[i * m + j];
                for k in 0..n {
                    if k != i {
                        sum -= a[i * n + k] * x[k * m + j];
                    }
                }
                x_new[i * m + j] = sum / diag;
            }
        }

        // Check convergence
        let max_diff = x.iter()
            .zip(x_new.iter())
            .map(|(x, y)| (x - y).abs())
            .fold(0.0f64, |a, b| a.max(b));

        x = x_new;

        if max_diff < 1e-8 {
            break;
        }
    }

    x
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_reservoir_step() {
        let mut reservoir = Reservoir::new(10, 50);
        let input = vec![0.5; 10];
        let state = reservoir.step(&input);
        assert_eq!(state.len(), 50);
    }

    #[test]
    fn test_reservoir_forward() {
        let mut reservoir = Reservoir::new(5, 30);
        let sequence = vec![vec![0.1; 5]; 10];
        let states = reservoir.forward(&sequence);
        assert_eq!(states.len(), 10);
        assert_eq!(states[0].len(), 30);
    }

    #[test]
    fn test_reservoir_reset() {
        let mut reservoir = Reservoir::new(8, 40);
        reservoir.step(&vec![1.0; 8]);
        reservoir.reset();
        assert!(reservoir.state.iter().all(|&s| s == 0.0));
    }
}
